solidity/random-beacon/contracts/libraries/Relay.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+// TODO: Documentation
+// TODO: Unit tests
+library Relay {
+    struct Request {
+        uint256 id;
+        uint256 startBlock;
+        uint256 groupIndex;
+        bytes previousEntry;
+    }
+
+    struct Data {
+        uint256 requestCount;
+        Request currentRequest;
+        uint256 groupSize;
+        uint256 entrySubmissionBlockStep;
+        uint256 bleedingPeriod;
+    }
+
+    function requestEntry(
+        Data storage self,
+        uint256 groupIndex,
+        bytes calldata previousEntry
+    ) internal {
+        require(!isRequestInProgress(self), "Another relay request in progress");
+
+        self.requestCount++;
+
+        self.currentRequest = Request(
+            self.requestCount,
+            block.number,
+            groupIndex,
+            previousEntry
+        );
+
+        // TODO: Emit event.
```

Comment:
Bear in mind events will have to be re-declared in the contract.

Reply:
Yep, I'm aware and I'm thinking about emitting events just from the contract and not from the library at all. WDYT?